### ISSUE TITLE
Copying default values from configspec changes encoding

### ISSUE THREAD

**Issue by dryo** [DiffSK/configobj#86]
When specifying an encoding for a config file with a configspec applying the configspec to copy the default values will change the encoding of the config file.

Here is the python snippet that shows the problem:

    from configobj import ConfigObj
    from validate import Validator

    configspec = '''
        version = float
    '''.splitlines()

    config = ConfigObj(configspec=configspec,
                       write_empty_values=True,
                       create_empty=True,
                       encoding='utf8')

    # verify that encoding is set correctly:
    print(config.encoding)  # will return 'utf8', correct

    # apply configspec while copying default values
    config.validate(Validator(), copy=True)

    # again verify that encoding is set correctly:
    print(config.encoding)  # will return 'None', incorrect

To get around this you have to create the configspec as a distinct ConfigObject object specifying its encoding and hand it over to the real ConfigObj.

**Comment by EliAndrewC** [comment]
Hmm, interesting.  I've confirmed the bug on my own machine, and it appears that this line is the culprit:
https://github.com/DiffSK/configobj/blob/0daebcb55aec2b5a18b287bbb2f4f6f03599f00e/configobj.py#L2182

In cases such as this where the configspec has no encoding set and the ``ConfigObj`` does, it would be easy enough to default to the ``ConfigObj`` encoding.  In cases where the two have explicitly set different encodings I'm less sure of the correct approach.  If not for backwards-compatibility concerns I'd just raise an exception if they don't match, because there's a lot of room for subtle encoding bugs in that case.  As it is, I'm still be tempted to do that, but probably can't since I imagine it would break "working" code in the wild (I put that in quotes since a lot of that code might very well have bugs that just haven't come up yet).

Given the difficulties in doing re-encoding, I'm inclined to just replace the line with something like

```python
if section.configspec.encoding is not None:
    section.encoding = section.configspec.encoding
```

**Comment by dryo** [comment]
I agree that it's most convenient (and also I don't think anyone would do it another way) if configspec and config file have the same encoding. But when an encoding is specifically set for the config file then IMO it's wrong to just change it. So I would propose the following:

  - If config file and configspec both have an encoding other than 'None' and it's different: raise an error
  - If config file and configspec both have an encoding other than 'None' and it's the same: do nothing
  - If config file has an encoding other than 'None' and configspec has the encoding 'None': copy encoding from config file to the configspec
  - If config file has the encoding 'None' and configspec has an encoding other than 'None': copy encoding from configspec to config file
  - If config file and configspec both have the encoding 'None': do nothing

I think this is a logic and clean approach that hopefully should not break any existing code.

**Comment by EliAndrewC** [comment]
After thinking about it for awhile I think you're right - the approach you suggest is unlikely to break any existing code and seems like the safest (i.e. least likely to result in a weird encoding bug) approach.

A pull request would be welcome, but if you don't have time then I'll make sure this is done before we tag the next release, whenever that ends up happening.

**Comment by jhermann** [comment]
Actually all issues in the 5.1.0 milestone are done.

Know what I mean? Nudge, nudge. Wink wink. :wink:

**Comment by EliAndrewC** [comment]
True enough!  I've been meaning to try to cut the 5.1.0 release for a few months now but just haven't gotten around to it.  My work schedule has simmered down lately so I should be able to find the time for it next week (fingers crossed).
